src/os/LinuxPWM.hxx

```diff
@@ -0,0 +1,174 @@
+ * @date 19 Feburary 2019
+ */
+
+#ifndef __LINUXPWM_HXX
+#define __LINUXPWM_HXX
+
+#include "freertos_drivers/common/PWM.hxx"
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <fcntl.h>
+#include <stdio.h>
+#include <string.h>
+
+class LinuxPWM : public PWM
+{
+public:
+    LinuxPWM(int chip, int channel) : chip_(chip) , channel_(channel)
+    {
+        snprintf(pwmdir,sizeof(pwmdir),
+                 "/sys/class/pwm/pwmchip%d/pwm%d",
+                 chip_,channel_);
+    }
+    void exportPin()
+    {
+        char exportname[50];
+        strcpy(exportname,pwmdir);
+        strcat(exportname,"/period");
+        FILE *fp = fopen(exportname,"r");
+        if (fp != NULL) {
+            fclose(fp);
+            set_period(1);
+            return;
+        }
+        char *p = strrchr(exportname,'/');
+        *p = '\0';
+        p = strrchr(exportname,'/');
+        p++;
+        strcpy(p,"export");
+        fp = fopen(exportname,"w");
+        fprintf(fp,"%d\n",channel_);
+        fclose(fp);
+        set_period(1);
+    }       
+    void set_period(uint32_t counts) override
+    {
+        char periodname[60];
+        strcpy(periodname,pwmdir);
+        strcat(periodname,"/period");
+        FILE *fp = fopen(periodname,"w");
+        fprintf(fp,"%d\n",counts);
+        fclose(fp);
+        if (counts > 0) {
+            enable();
+        } else {
+            disable();
+        }
+    }
+    uint32_t get_period() override
+    {
+        char periodname[60];
+        uint32_t counts;
+        strcpy(periodname,pwmdir);
+        strcat(periodname,"/period");
+        FILE *fp = fopen(periodname,"r");
+        fscanf(fp,"%d",&counts);
+        fclose(fp);
+        return counts;
+    }
+    void set_duty(uint32_t counts) override
+    {
+        char duty_cyclename[60];
+        strcpy(duty_cyclename,pwmdir);
+        strcat(duty_cyclename,"/duty_cycle");
+        FILE *fp = fopen(duty_cyclename,"w");
+        fprintf(fp,"%d\n",counts);
+        fclose(fp);
+        if (counts > 0) {
+            enable();
+        } else {
+            disable();
+        }
+    }
+    uint32_t get_duty() override
+    {
+        char duty_cyclename[60];
+        uint32_t counts;
+        strcpy(duty_cyclename,pwmdir);
+        strcat(duty_cyclename,"/duty_cycle");
+        FILE *fp = fopen(duty_cyclename,"r");
+        fscanf(fp,"%d",&counts);
+        fclose(fp);
+        return counts;
+    }
+    uint32_t get_period_max() override
+    {
+        return 0xffffffff;
+    }
+    uint32_t get_period_min() override
+    {
+        return 1;
+    }
+    void enable() {
+        char enablename[60];
+        strcpy(enablename,pwmdir);
+        strcat(enablename,"/enable");
+        int fd = open(enablename,O_WRONLY);
+        write(fd,"1\n",2);
+        close(fd);
+    }
+    void disable() {
+        char enablename[60];
+        strcpy(enablename,pwmdir);
+        strcat(enablename,"/enable");
+        int fd = open(enablename,O_WRONLY);
```

Comment:
is there a reason you opted for mixing fopen and open? it might make more sense to use just one api.

Reply:
Only that when I am writing a pure constant (eg for the enable/disable functions), I don't need the overhead of fprintf().  But where I need to convert a number to a string, I need fprintf() (or snprintf() if I used open()...).  Why? Does it really matter?  Remember, this is on machines running a fullfledged OS (Linux), not on MCUs running a minimualist OS (FreeRTOS) out of Flash memorty.

Reply:
Fair enough and makes some sense. There may even be other options available in OpenMRN that I haven't explored yet as well.